usr.bin/cs/cs.1
.\"
.\" Copyright (c) 2022 Edward Tomasz Napierala <en322@cl.cam.ac.uk>
.\" All rights reserved.
.\"
.\" This software was developed by the University of Cambridge Computer
.\" Laboratory as part of the CHERI for Hypervisors and Operating Systems
.\" (CHaOS) project, funded by EPSRC grant EP/V000292/1.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD$
.\"
.Dd May 25, 2022
.Dt CS 1
.Os
.Sh NAME
.Nm cs
.Nd Capsicum service
.Sh SYNOPSIS
.Nm
.Op Fl k
.Op Fl q
.Op Fl v
.Ar command Op Ar args ...
.Sh DESCRIPTION
The
.Nm
utility is used to transparently force its descendant into
.Xr capsicum 4
capability mode, and to provide it with remote syscall service.
It spawns a child process, makes it preload a wrapper which replaces
every system call that would otherwise be denied by the
.Xr capsicum 4
framework with a
.Xr cocall 2
to
.Nm
service, and finally executes
.Ar command
with the sealed capability to
.Nm
service inserted into its inherited capability vector.
In the child process, the preloaded wrapper then calls
.Xr cap_enter 2
in its constructor, making the process enter capability mode.
.Pp
The following options are available:
.Bl -tag -width ".Fl r Ar path"
.It Fl k
Use
.Xr coaccept_slow 2
instead of
.Xr coaccept 2 .
.It Fl q
Be quiet; do not print out the file paths it denied access to.
.It Fl r Ar path
Make the
.Ar path
readable to the child process.
.It Fl w Ar path
Make the
.Ar path
readable and writable to the child process.
.It Fl v
Be verbose.
.El
.Pp
.Nm
can be used as a template when implementing new
.Xr cocall 2
services that require file descriptor passing using
.Xr capfromfd 2
and
.Xr captofd 2 .
.Sh EXIT STATUS
The
.Nm
utility exits 0 on success, and >0 if an error occurs.
.Sh EXAMPLES
Run
.Xr ls 1
to list files in
.Pa /usr
without giving it access to any other files:
.Dl cs -r /usr ls -al /usr
.Pp
See what changes when you also allow
.Pa /etc
and silence down further warnings:
.Dl cs -qr /usr -r /etc ls -al /usr
.Sh SEE ALSO
.Xr capv 1 ,
.Xr coaccept 2 ,
.Xr capsicum 4 ,
.Xr colocation 7
.Sh BUGS
.Nm
cannot protect against malicious binaries - only against
otherwise trustworthy binaries becoming compromised by malicious data.
While
.Ar command
does eventually end up running in
.Xr capsicum 4
mode, it is only entered from an LD_PRELOADed constructor, just before starting the
.Fn main
function, and can be easily
disabled by a malicious executable.
To make it protect against malicious binaries, they would need to be started in
.Xr capsicum 4
mode outright.
This would require implementing syscall remoting also for
.Xr rtld 1 ,
precluding the use of LD_PRELOAD.
It would probably require moving the stubs directly into libc and rtld.
.Pp
Because after
.Xr fork 2
the child process is running in a fresh new address space
and thus unable to communicate with colocated services,
and because
.Xr vfork 2
is not currently allowed in
.Xr capsicum 4
capability mode, it is not possible for
.Nm
to support more than a single client process at a time.
.Pp
If you allow access to
.Pa /proc
you'll then be able to open
.Pa /proc/curproc/mem
and take over the
.Nm
instance.
If you allow access to
.Pa /dev ,
you will be able to take over the whole machine.
So maybe don't.
In the long run,
instead of calling the syscalls on behalf of the client,
.Nm
should be returning file descriptors to allowed directories (specified with
.Fl r )
for the clients to use them to call syscalls themselves.
.Pp
.Nm
does not support any kind of path translation.
This is a deliberate omission, to avoid mixing resource mapping
with access control like NAT did.
.Sh HISTORY
The
.Nm
command appeared in
.Tn CheriBSD .
.Sh AUTHORS
.An -nosplit
The
.Nm
utility was developed by
.An Edward Tomasz Napierala Aq Mt trasz@FreeBSD.org
at the University of Cambridge Computer Laboratory as part of the CHERI
for Hypervisors and Operating Systems (CHaOS) project, funded by EPSRC
grant EP/V000292/1.


usr.bin/cs/cs.c
/*-
 * Copyright (c) 2022 Edward Tomasz Napierala <trasz@FreeBSD.org>
 * All rights reserved.
 *
 * This software was developed by the University of Cambridge Computer
 * Laboratory as part of the CHERI for Hypervisors and Operating Systems
 * (CHaOS) project, funded by EPSRC grant EP/V000292/1.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * This is the callee (coaccepting) part of the syscall remoting service.
 * with the client (cocalling) counterpart in lib/libcs/cs.c.
 *
 * At this point it just directly executes the syscalls it receives.
 * Implementing an actual policy (or telling the client processes apart)
 * is left as an exercise for the reader.
 */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD$");

#include <sys/auxv.h>
#include <sys/capsicum.h>
#include <sys/param.h>
#include <sys/queue.h>
#include <sys/sbuf.h>
#include <sys/syscall.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <assert.h>
#include <capv.h>
#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <signal.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sysdecode.h>
#include <unistd.h>

struct pathname {
	LIST_ENTRY(pathname)	p_link;
	int			p_fd;
	char			*p_path;
};

LIST_HEAD(, pathname) pathnames = LIST_HEAD_INITIALIZER(pathnames);
static bool kflag = false, qflag = false, vflag = false;

static void
usage(void)
{

	fprintf(stderr, "usage: cs [-kqv] [-r path] [-w path] command [args ...]\n");
	exit(0);
}

static void
sigchld_handler(int dummy __unused)
{

	exit(0);
}

static void
answerback(capv_answerback_t *out)
{
	struct sbuf sb;
	const struct pathname *r;
	int error;

	memset(out, 0, sizeof(*out));
	out->len = sizeof(*out);
	out->op = 0;
	sbuf_new(&sb, out->answerback, sizeof(out->answerback), SBUF_FIXEDLEN);

	sbuf_printf(&sb, "cs");
	LIST_FOREACH(r, &pathnames, p_link)
		sbuf_printf(&sb, " -r %s", r->p_path);
	sbuf_printf(&sb, ", pid %d", getpid());
	if (kflag)
		sbuf_printf(&sb, " (slow)");

	error = sbuf_finish(&sb);
	if (error != 0)
		err(1, "sbuf_finish");
}

static void
capvreturn(capv_syscall_return_t *out, int op, int error, int errno_, uintcap_t fdcap)
{
	memset(out, 0, sizeof(*out));
	out->len = sizeof(*out);
	out->op = op;
	out->error = error;
	out->errno_ = errno_;
	out->fdcap = fdcap;
}

static void
add_pathname(const char *path)
{
	struct pathname *r;

	r = calloc(1, sizeof(*r));
	if (r == NULL)
		err(1, "calloc");
	r->p_path = realpath(path, NULL);
	if (r->p_path == NULL)
		err(1, "realpath %s", path);
	r->p_fd = open(r->p_path, O_RDONLY);
	if (r->p_fd < 0)
		err(1, "%s", r->p_path);
	LIST_INSERT_HEAD(&pathnames, r, p_link);
}

/*
 * XXX: If this whole function makes you feel uncomfortable, that's a good sign :)
 * This needs to be replaced by returning p_fd to the calling process,
 * to pass as fd to whateverat(2) on their side instead of doing it here...
 * but that has its own problems.
 */
static int
handle_pathname(int fd, const char *path)
{
	const struct pathname *r;
	char resolved[PATH_MAX];
	int error;

#ifdef notyet
	error = realpathat(fd, path, resolved, sizeof(resolved), 0);
#else
	/*
	 * XXX: Not exported from libc for some reason.
	 */
	error = syscall(SYS___realpathat, fd, path, resolved, sizeof(resolved), 0);
#endif
	if (error < 0) {
		if (!qflag)
			warn("__realpathat(%d, \"%s\", ...)", fd, path);
		return (EPERM);
	}

	LIST_FOREACH(r, &pathnames, p_link) {
		if (strncmp(resolved, r->p_path, strlen(r->p_path)) == 0)
			return (0);
	}

	return (EPERM);
}

int
main(int argc, char **argv)
{
	capv_syscall_t in;
	union {
		capv_answerback_t answerback;
		capv_syscall_return_t syscall;
	} outbuf;
	capv_syscall_return_t *out = &outbuf.syscall;
	struct sigaction sa;
	void * __capability public;
	void * __capability cookie;
	void * __capability *capv = NULL;
	char *ld_preload;
	char *tmp = NULL;
	const char *path;
	uintcap_t fdcap;
	ssize_t received;
	pid_t pid;
	int capc, ch, error, _error, fd;

	LIST_INIT(&pathnames);

	while ((ch = getopt(argc, argv, "kqr:w:v")) != -1) {
		switch (ch) {
		case 'k':
			kflag = true;
			break;
		case 'q':
			qflag = true;
			break;
		case 'r':
			add_pathname(optarg);
			break;
		case 'v':
			vflag = true;
			break;
		case '?':
		default:
			usage();
		}
	}

	argc -= optind;
	argv += optind;
	if (argc < 1)
		usage();

	memset(&sa, 0, sizeof(sa));
	sa.sa_handler = sigchld_handler;
	sa.sa_flags = SA_NOCLDSTOP;
	sigfillset(&sa.sa_mask);

	/*
	 * We are spawning child processes, so we'll need to handle SIGCHLD.
	 */
	error = sigaction(SIGCHLD, &sa, NULL);
	if (error != 0)
		err(1, "sigaction");

	error = cosetup(COSETUP_COACCEPT);
	if (error != 0)
		err(1, "cosetup");

	error = coregister(NULL, &public);
	if (error != 0)
		err(1, "coregister");

	capvfetch(&capc, &capv);
	error = capvset(&capc, &capv, CAPV_SYSCALL, public);
	if (error != 0)
		err(1, "capvset");

	/*
	 * This whole mess with environment variables is to preload
	 * libcs.so, which provides replacement stubs to shadow
	 * the system calls.  This makes unmodified purecap binaries
	 * transparently call cs(1) instead of kernel.  A bit ugly,
	 * but convenient for now.
	 */
	ld_preload = getenv("LD_PRELOAD");
	if (ld_preload != NULL) {
		asprintf(&tmp, "%s:%s", ld_preload, "/usr/lib/libcs.so");
	} else {
		asprintf(&tmp, "%s", "/usr/lib/libcs.so");
	}
	error = setenv("LD_PRELOAD", tmp, 1);
	if (error != 0)
		err(1, "setenv");

	/*
	 * We can't explicitly pass capv into another address space,
	 * so we need vfork(2) here, not fork(2).
	 */
	pid = vfork();
	if (pid < 0)
		err(1, "vfork");

	if (pid == 0) {
		/*
		 * Child, will coexecvec(2) the new command.
		 */
		coexecvpc(getppid(), argv[0], argv, capv, capc);

		/*
		 * Shouldn't have returned.
		 */
		err(1, "%s", argv[0]);
	}

	/*
	 * Parent, will loop on coaccept(2) until SIGCHLD.
	 */

	memset(out, 0, sizeof(*out));

	for (;;) {
		/*
		 * Send back a response, if any, and wait for the next call.
		 */
		if (kflag)
			received = coaccept_slow(&cookie, out, out->len, &in, sizeof(in));
		else
			received = coaccept(&cookie, out, out->len, &in, sizeof(in));
		if (received < 0) {
			warn("%s", kflag ? "coaccept_slow" : "coaccept");
			memset(out, 0, sizeof(*out));
			continue;
		}

		/*
		 * Answered.
		 */
		if (vflag) {
			error = cocachedpid(&pid, cookie);
			if (error != 0)
				warn("cogetpid");
			printf("%s: op %d<%s>, len %zd from pid %d -> pid %d%s\n",
			    getprogname(), in.op, sysdecode_syscallname(SYSDECODE_ABI_FREEBSD, in.op),
			    in.len, pid, getpid(), kflag ? " (slow)" : "");
		}

		/*
		 * Many syscalls are similar to one another; handle the similarities
		 * here.
		 */
		error = errno = fdcap = 0;

		switch (in.op) {
		case 0:
			/*
			 * Is this a proper packet?  Op 0 is answerback request, so the size
			 * obviously won't match; just make sure that we've received the op field.
			 */
			if ((size_t)received < sizeof(capv_t)) {
				warnx("size mismatch: received %zd, expected %zd",
				    (size_t)received, sizeof(capv_t));
				capvreturn(out, 0, -1, ENOMSG, 0);
				goto respond;
			}

			answerback(&outbuf.answerback);
			goto respond;
		case SYS_bind:
		case SYS_connect:
			/*
			 * Is this a proper packet?
			 */
			if ((size_t)received != sizeof(in)) {
				warnx("size mismatch: received %zd, expected %zd",
				    (size_t)received, sizeof(in));
				capvreturn(out, 0, -1, ENOMSG, 0);
				goto respond;
			}

			/*
			 * Receive the socket descriptor.
			 */
			error = captofd((void *)in.arg[0], (int *)&in.arg[0]);
			if (error != 0) {
				warn("captofd: %#lp", (void *)in.arg[0]);
				capvreturn(out, -in.op, error, ENOMSG, 0);
				goto respond;
			}
			break;
		case SYS_fchdir:
			/*
			 * Syscalls that take an fd for their first argument,
			 * and there's no path argument.
			 */
			if ((size_t)received != sizeof(in)) {
				warnx("size mismatch: received %zd, expected %zd",
				    (size_t)received, sizeof(in));
				capvreturn(out, 0, -1, ENOMSG, 0);
				goto respond;
			}

			error = captofd((void *)in.arg[0], (int *)&in.arg[0]);
			if (error != 0) {
				warn("captofd: %#lp", (void *)in.arg[0]);
				capvreturn(out, -in.op, error, ENOMSG, 0);
				goto respond;
			}
			break;
		case SYS_fstatat:
		case SYS_fchmodat:
		case SYS_fchownat:
		case SYS_utimensat:
		case SYS_openat:
			/*
			 * Syscalls that take an fd for their first argument,
			 * and a path for the second.
			 */
			if ((size_t)received != sizeof(in)) {
				warnx("size mismatch: received %zd, expected %zd",
				    (size_t)received, sizeof(in));
				capvreturn(out, 0, -1, ENOMSG, 0);
				goto respond;
			}

			if (in.arg[0] == (uintcap_t)AT_FDCWD) {
				fd = AT_FDCWD;
			} else {
				error = captofd((void *)in.arg[0], (int *)&in.arg[0]);
				if (error != 0) {
					warn("captofd: %#lp", (void *)in.arg[0]);
					capvreturn(out, -in.op, error, ENOMSG, 0);
					goto respond;
				}
				fd = (int)in.arg[0];
			}

			path = (const char *)in.arg[1];
			error = handle_pathname(fd, path);
			if (error != 0) {
				if (!qflag)
					warnx("refusing to %s %s", sysdecode_syscallname(SYSDECODE_ABI_FREEBSD, in.op), path);
				capvreturn(out, -in.op, -1, error, 0);
				goto respond;
			}
			break;
		case SYS_mkdir:
		case SYS_rmdir:
		case SYS_pathconf:
		case SYS_lpathconf:
		case SYS___getcwd:
			/*
			 * Syscalls that take a path as their first argument.
			 */
			if ((size_t)received != sizeof(in)) {
				warnx("size mismatch: received %zd, expected %zd",
				    (size_t)received, sizeof(in));
				capvreturn(out, 0, -1, ENOMSG, 0);
				goto respond;
			}

			path = (const char *)in.arg[0];
			error = handle_pathname(AT_FDCWD, path);
			if (error != 0) {
				warnx("refusing to %s %s", sysdecode_syscallname(SYSDECODE_ABI_FREEBSD, in.op), path);
				capvreturn(out, -in.op, -1, error, 0);
				goto respond;
			}
			break;
		default:
			warnx("unknown op %d<%s>", in.op, sysdecode_syscallname(SYSDECODE_ABI_FREEBSD, in.op));
			capvreturn(out, -in.op, -1, ENOSYS, 0);
			goto respond;
		}

		/*
		 * Do the thing.
		 */
		error = syscall(in.op, in.arg[0], in.arg[1], in.arg[2], in.arg[3], in.arg[4], in.arg[5], in.arg[6]);
		if (vflag && error != 0 && errno != 0) /* Check errno to not display this warning on every openat(2) */
			warn("syscall(%d<%s>, ...)", in.op, sysdecode_syscallname(SYSDECODE_ABI_FREEBSD, in.op));

		/*
		 * Any file descriptors to return?
		 */
		switch (in.op) {
		case SYS_openat:
			/*
			 * NB: It's actually a file descriptor number, not an error.
			 */
			if (error > 0) {
				_error = capfromfd((void *)&fdcap, error);
				if (_error != 0) {
					warn("capfromfd");
					capvreturn(out, -in.op, error, ENOMSG, 0);
					goto respond;
				}
			}
			break;
		}

		/*
		 * Cleanup, again supposed to be shared between multiple syscalls.
		 */
		switch (in.op) {
		case SYS_bind:
		case SYS_connect:
			_error = close((int)in.arg[0]);
			if (_error != 0)
				warn("close(%d)", (int)in.arg[0]);
			break;
		case SYS_fstatat:
		case SYS_fchmodat:
		case SYS_fchownat:
		case SYS_utimensat:
		case SYS_openat:
		case SYS_fchdir:
			if (in.arg[0] != (uintcap_t)AT_FDCWD) {
				_error = close((int)in.arg[0]);
				if (_error != 0)
					warn("close(%d)", (int)in.arg[0]);
			}
			break;
		}

		/*
		 * XXX: So when do we close the descriptor returned by openat?
		 */

		capvreturn(out, -in.op, error, errno, fdcap);

respond:
		/*
		 * Send the response back and loop.
		 */
		if (vflag) {
			printf("%s: returning to pid %d <- pid %d: op %d, len %zd, error %d, errno %d%s\n",
			    getprogname(), pid, getpid(), out->op, out->len, out->error, out->errno_, kflag ? " (slow)" : "");
		}
	}
}
